config/compression.go

```diff
@@ -0,0 +1,84 @@
+package config
+
+import (
+	"fmt"
+	"strings"
+)
+
+type Compression struct {
+	Request  ReqCompression  `mapstructure:"request"`
+	Response RespCompression `mapstructure:"response"`
+}
+
+// CompressionConfig defines if compression is enabled and what type of compression to use
+type ReqCompression struct {
+	Enabled bool     `mapstructure:"enabled"`
+	Kind    []string `mapstructure:"kind,flow"`
+	kindMap map[CompressionKind]bool
+}
+
+type RespCompression struct {
+	Enabled bool   `mapstructure:"enabled"`
+	Kind    string `mapstructure:"kind"`
+}
+
+type CompressionKind string
+
+const (
+	CompressionGZIP CompressionKind = "gzip"
+)
+
+func (k CompressionKind) IsValid() bool {
+	switch k {
+	// Case for valid types. As new compression types are added they should
+	// be added here as a comma separated list.
+	case CompressionGZIP:
+		return true
+	default:
+		return false
+	}
+}
+
+func (cfg *Compression) validate(errs []error) []error {
+	errs = cfg.Request.Validate(errs)
+	errs = cfg.Response.Validate(errs)
+	return errs
+}
+
+func (cfg *ReqCompression) IsSupported(kind CompressionKind) bool {
+	if cfg.Enabled {
+		return cfg.kindMap[kind]
+	}
+
+	return false
+}
+
+func (cfg *ReqCompression) Validate(errs []error) []error {
+	if cfg.Enabled {
+		if len(cfg.Kind) == 0 {
+			errs = append(errs, fmt.Errorf("compression is enabled but no compression types are specified"))
+		}
+
+		// This is to enabled O(1) lookups for supported compression types
+		cfg.kindMap = make(map[CompressionKind]bool, len(cfg.Kind))
+		for _, kind := range cfg.Kind {
+			k := CompressionKind(strings.ToLower(kind))
+			if !k.IsValid() {
```

Comment:
Since `kind.IsValid()` is also acceptable. How about updating `IsValid` to convert `k` into lowercase?  This change safeguards us from error when caller function does not  convert into kind into lowercase.   
```
func (k CompressionKind) IsValid() bool {
	- switch k {
	+ switch strings.ToLower(kind) {
	// Case for valid types. As new compression types are added they should
	// be added here as a comma separated list.
	case CompressionGZIP:
		return true
	default:
		return false
	}
}
```

Reply:
We can't call `strings.ToLower(k)` because `k` is of type `CompressionKind` and I thought about this but didn't do it because in the `Validate()` method we also create the `kindMap` and we want to add values that are converted to lower case too in there so that `IsSupported()` doesn't need to do the conversion again. And the other reason was switching the values back and forth between `CompressionKind` and `string` to make this work. However, I think we can take care of that by just making `Kind` to be a `CompressionKind` type rather than a `string`. This will still require us to convert to lower case in both `IsValid()` and `IsSupported()` but I agree that not needing to rely on the caller to do this does make it worth it.